### PR TITLE
shuf: move from getopts to clap

### LINKED ISSUE #1933: shuf: move from getopts to clap
Code is in:
https://github.com/uutils/coreutils/blob/master/src/uu/shuf/src/shuf.rs

Tests:
https://github.com/uutils/coreutils/blob/master/tests/by-util/test_shuf.rs 

As example:
https://github.com/uutils/coreutils/commit/9081e120d6b866c94679d9e29c7bdb17ccccb4d0

### PR DESCRIPTION
**Pull request by siyopao**
Closes #1933.
Part of #1859.

In a couple of places I updated the error message so that it is the same as the GNU version.

There are incompatible flags and I am wondering if it is better to use claps `conflicts_with` (current situation) or to check after the fact. In the latter case we can then use the same error message as the GNU version. I left a comment to this effect in the code and that should be cleaned up no matter which way you decide to go.

There is also the question of how closely you would like to follow the GNU versions help message. At the moment we have
```
$ ./target/debug/coreutils shuf --help
Usage: shuf [OPTION]... [FILE]
  or:  shuf -e [OPTION]... [ARG]...
  or:  shuf -i LO-HI [OPTION]...
Write a random permutation of the input lines to standard output.

With no FILE, or when FILE is -, read standard input.

Mandatory arguments to long options are mandatory for short options too.
    -e, --echo <ARG>...           treat each ARG as an input line
    -h, --help                    Prints help information
    -n, --head-count <COUNT>      output at most COUNT lines
    -i, --input-range <LO-HI>     treat each number LO through HI as an input line
    -o, --output <FILE>           write result to FILE instead of standard output
        --random-source <FILE>    get random bytes from FILE
    -r, --repeat                  output lines can be repeated
    -V, --version                 Prints version information
    -z, --zero-terminated         line delimiter is NUL, not newline
```
and GNUs version is
```
$ shuf --help
Usage: shuf [OPTION]... [FILE]
  or:  shuf -e [OPTION]... [ARG]...
  or:  shuf -i LO-HI [OPTION]...
Write a random permutation of the input lines to standard output.

With no FILE, or when FILE is -, read standard input.

Mandatory arguments to long options are mandatory for short options too.
  -e, --echo                treat each ARG as an input line
  -i, --input-range=LO-HI   treat each number LO through HI as an input line
  -n, --head-count=COUNT    output at most COUNT lines
  -o, --output=FILE         write result to FILE instead of standard output
      --random-source=FILE  get random bytes from FILE
  -r, --repeat              output lines can be repeated
  -z, --zero-terminated     line delimiter is NUL, not newline
      --help     display this help and exit
      --version  output version information and exit

GNU coreutils online help: <https://www.gnu.org/software/coreutils/>
Report any translation bugs to <https://translationproject.org/team/>
Full documentation <https://www.gnu.org/software/coreutils/shuf>
or available locally via: info '(coreutils) shuf invocation'
```

### CODE REVIEW

**Comment by sylvestre** [comment]
No need to touch the help, it is fine :)

File: src/uu/shuf/src/shuf.rs [hunk 1]
```diff
@@ -10,132 +10,165 @@
...
 }
 
 static NAME: &str = "shuf";
 static VERSION: &str = env!("CARGO_PKG_VERSION");
+static OPT_ECHO: &str = "echo";
+static OPT_INPUT_RANGE: &str = "input-range";
+static OPT_HEAD_COUNT: &str = "head-count";
+static OPT_OUTPUT: &str = "output";
+static OPT_RANDOM_SOURCE: &str = "random-source";
+static OPT_REPEAT: &str = "repeat";
```
Comment (ycd) [inline src/uu/shuf/src/shuf.rs:39]: Please use an options mod for this:
https://github.com/uutils/coreutils/blob/83f8140aafe80a9683eb008362caedef87edd831/src/uu/cat/src/cat.rs#L129-L140

File: src/uu/shuf/src/shuf.rs [hunk 2]
```diff
@@ -10,132 +10,165 @@
...
-    if matches.opt_present("help") {
-        let msg = format!(
-            "{0} {1}
+pub fn uumain(args: impl uucore::Args) -> i32 {
+    let matches = App::new(executable!())
+        .name(NAME)
+        .version(VERSION)
+        .template("Usage: {usage}\nMandatory arguments to long options are mandatory for short options too.\n{unified}")
+        .usage(USAGE)
+        .help_short("")
```
Comment (ycd) [inline src/uu/shuf/src/shuf.rs:56]: Setting this as empty will override the short auto-generated help.

**Review by ycd — CHANGES_REQUESTED** [review]

File: src/uu/shuf/src/shuf.rs (hunk 2 above)
Comment (ycd) [reply src/uu/shuf/src/shuf.rs:56]: Ah, i see now there is no `-h` for shuf :)

File: src/uu/shuf/src/shuf.rs [hunk 3]
```diff
@@ -10,132 +10,165 @@
...
-            }
-            Mode::Default => {
-                let fdata = read_input_file(&matches.free[0][..]);
-                let mut fdata = vec![&fdata[..]];
-                find_seps(&mut fdata, sep);
-                shuf_bytes(&mut fdata, repeat, count, sep, output, random);
-            }
+    match mode {
+        Mode::Echo(args) => {
+            // XXX: this doesn't correctly handle non-UTF-8 cmdline args
```
Comment (sylvestre) [inline src/uu/shuf/src/shuf.rs:157]: is this comment still relevant with clap?

**Review by sylvestre — CHANGES_REQUESTED** [review]

File: src/uu/shuf/src/shuf.rs (hunk 2 above)
Comment (siyopao) [reply src/uu/shuf/src/shuf.rs:56]: Oops, this didn't actually work and I forgot to remove it. I'll remove it. Thanks!

File: src/uu/shuf/src/shuf.rs (hunk 1 above)
Comment (siyopao) [reply src/uu/shuf/src/shuf.rs:39]: Sure! Thanks.

File: src/uu/shuf/src/shuf.rs [hunk 4]
```diff
@@ -10,132 +10,165 @@
...
-    Default,
-    Echo,
+    Default(String),
+    Echo(Vec<String>),
     InputRange((usize, usize)),
 }
 
 static NAME: &str = "shuf";
 static VERSION: &str = env!("CARGO_PKG_VERSION");
+static OPT_ECHO: &str = "echo";
```
[the remark this answers is not in the document]
Comment (siyopao) [reply src/uu/shuf/src/shuf.rs:34]: Okay. Thanks!

File: src/uu/shuf/src/shuf.rs [hunk 5]
```diff
@@ -10,132 +10,165 @@
...
+                .help("line delimiter is NUL, not newline"),
+        )
+        .arg(Arg::with_name(OPT_FILE).takes_value(true))
+        .get_matches_from(args);
 
-Usage:
-  {0} [OPTION]... [FILE]
-  {0} -e [OPTION]... [ARG]...
-  {0} -i LO-HI [OPTION]...
+    // TODO: Which is better. To catch it using `conflicts_with` above or here so we can have the same error message as the GNU version?
```
[the remark this answers is not in the document]
Comment (siyopao) [reply src/uu/shuf/src/shuf.rs:116]: Thank you. Comment removed.

File: src/uu/shuf/src/shuf.rs (hunk 3 above)
Comment (siyopao) [reply src/uu/shuf/src/shuf.rs:157]:
```
$ ./target/debug/coreutils shuf -e こんにちは -e t
t
こんにちは
```
No, it seems like it is no longer relevant. I'll remove it. Thanks!

File: Cargo.lock [hunk 6]
```diff
@@ -1,1434 +1,1591 @@
 # This file is automatically @generated by Cargo.
 # It is not intended for manual editing.
+version = 3
```
Comment (sylvestre) [inline Cargo.lock:3]: please do it this way:
cargo +1.40.0 update

File: src/uu/shuf/src/shuf.rs [hunk 7]
```diff
@@ -10,132 +10,161 @@
...
+    pub static ZERO_TERMINATED: &str = "zero-terminated";
+    pub static FILE: &str = "file";
+}
 
 pub fn uumain(args: impl uucore::Args) -> i32 {
-    let args = args.collect_str();
+    let matches = App::new(executable!())
+        .name(NAME)
+        .version(VERSION)
+        .template("Usage: {usage}\nMandatory arguments to long options are mandatory for short options too.\n{unified}")
```
Comment (sylvestre) [inline src/uu/shuf/src/shuf.rs:57]: please declare it like usage

File: src/uu/shuf/src/shuf.rs [hunk 8]
```diff
@@ -10,132 +10,161 @@
...
-                        show_error!("extra operand '{}'", &matches.free[1][..]);
-                    }
-                    Mode::Default
-                }
-            }
-        };
-        let repeat = matches.opt_present("repeat");
-        let sep = if matches.opt_present("zero-terminated") {
+            },
+            None => usize::MAX,
```
Comment (sylvestre) [inline src/uu/shuf/src/shuf.rs:140]: this is failing with rust 1.40

File: tests/by-util/test_shuf.rs [hunk 9]
```diff
@@ -1 +1,55 @@
-// ToDO: add tests
+// ToDO: add more tests
```
Comment (sylvestre) [inline tests/by-util/test_shuf.rs:1]: kudos for the tests :)

it would be nice to also test success and not only failures ;)

**Review by sylvestre — CHANGES_REQUESTED** [review]

File: src/uu/shuf/src/shuf.rs (hunk 8 above)
Comment (siyopao) [reply src/uu/shuf/src/shuf.rs:140]: Ah, okay, I'll use `std::usize::MAX`.

File: tests/by-util/test_shuf.rs (hunk 9 above)
Comment (siyopao) [reply tests/by-util/test_shuf.rs:1]: I think @jaggededgedjustice said that they would like to do it in #1859, and I didn't want to step on any toes 😄

File: src/uu/shuf/src/shuf.rs (hunk 7 above)
Comment (siyopao) [reply src/uu/shuf/src/shuf.rs:57]: Right-ho. Done. Thanks.

File: Cargo.lock (hunk 6 above)
Comment (siyopao) [reply Cargo.lock]: Ah, thanks for that. I wasn't sure that `Cargo.lock` even belonged here, but the tests were failing without it.

File: tests/by-util/test_shuf.rs (hunk 9 above)
Comment (jaggededgedjustice) [reply tests/by-util/test_shuf.rs:1]: That's correct, tests will be along soon.

**Pull request merged**